metals/src/main/scala/scala/meta/internal/metals/ScalafixProvider.scala

```diff
@@ -64,64 +67,87 @@ case class ScalafixProvider(
-            scribe.warn(
-              msg
-            )
-            languageClient.showMessage(
-              MessageType.Warning,
-              msg
-            )
-            Future.successful(Nil)
-          case Success(results) if !scalafixSucceded(results) =>
-            val scalafixError = getMessageErrorFromScalafix(results)
-            val exception = ScalafixRunException(scalafixError)
-            reportScalafixError(
-              scalafixError,
-              exception
-            )
+
+    compilations.compilationFinished(file).flatMap { _ =>
+      val temporarySemanticdbDirectory =
+        if (retried || isUnsaved(inBuffers.text, fromDisk.text))
+          createTemporarySemanticdb(file)
+        else None
+      val scalafixEvaluation =
+        scalafixEvaluate(file, scalaTarget, temporarySemanticdbDirectory)
+
+      temporarySemanticdbDirectory.foreach { path =>
+        AbsolutePath(path).deleteRecursively
+      }
+
+      scalafixEvaluation match {
+        case Failure(exception) =>
+          reportScalafixError(
+            "Unable to run scalafix, please check logs for more info.",
+            exception
+          )
+          Future.failed(exception)
+        case Success(results)
+            if !scalafixSucceded(results) && hasStaleSemanticdb(
+              results
+            ) && buildClient.buildHasErrors(file) =>
+          val msg = "Attempt to organize your imports failed. " +
+            "It looks like you have compilation issues causing your semanticdb to be stale. " +
+            "Ensure everything is compiling and try again."
+          scribe.warn(
+            msg
+          )
+          languageClient.showMessage(
+            MessageType.Warning,
+            msg
+          )
+          Future.successful(Nil)
+        case Success(results) if !scalafixSucceded(results) =>
+          val scalafixError = getMessageErrorFromScalafix(results)
+          val exception = ScalafixRunException(scalafixError)
+          reportScalafixError(
+            scalafixError,
+            exception
+          )
+          if (!retried && hasStaleSemanticdb(results)) {
+            // Retry, since the semanticdb might be stale
+            organizeImports(file, scalaTarget, retried = true)
+          } else {
             Future.failed(exception)
-          case Success(results) =>
-            Future.successful {
-              val edits = for {
-                fileEvaluation <- results.getFileEvaluations().headOption
-                patches <- fileEvaluation.previewPatches().asScala
-              } yield textEditsFrom(patches, fromDisk)
-              edits.getOrElse(Nil)
-            }
-        }
+          }
+        case Success(results) =>
+          Future.successful {
+            val edits = for {
+              fileEvaluation <- results.getFileEvaluations().headOption
+              patches <- fileEvaluation.previewPatches().asScala
+            } yield textEditsFrom(patches, fromDisk)
```

Comment:
Won't this cause issues in the situation where the buffer doesn't match what was last saved on disk? For example if the user has an unsaved buffer and they trigger this, it will indeed now run the organize imports, but it also then applies those patches to the last saved file on disk resulting in reverting their code back. For example:

![2021-08-23 09 24 26](https://user-images.githubusercontent.com/13974112/130407210-d7eec385-3116-4189-8fcd-7a0f134610d6.gif)

Notice how it removed the `val u = word.toUpperCase()` that I just added.

Reply:
Ach, good catch, this should use the value from buffers. Thanks!

Reply:
That should work correctly now, this was also hiding the issue that Scalafix actually reads the file, so this could not have worked at all. I worked around it by saving both semanticdb file and the source file contained in the buffers to tmp.